QuarkusComponentTest does not register Quarkus Config Converters

### Describe the bug

A QuarkusComponentTest fails for components which rely on MicroProfile Config Converters provided by Quarkus. 
E.g. `io.quarkus.runtime.configuration.DurationConverter`.

Testing a component using
```
@ConfigProperty(name = "my.duration", defaultValue = "60s")
private Duration myDuration;
```
fails with an exception: `java.lang.RuntimeException: Error injecting java.time.Duration io.quarkus.issue.MyComponent.myDuration`.

Running the application works fine since then the Quarkus Config Converters are available.

### Expected behavior

Expected the Quarkus Config Converters to be available and used when running a QuarkusComponentTest test
or to have some kind of mechanism with which custom converters can be registered.

### Actual behavior

None of the Quarkus Config Converters are registered.

The exception is caused by the default Duration converter provided by SmallRye's `io.smallrye.config.ImplicitConverters` which tries to use the `java.time.Duration.parse` function to parse the value but this does not support the syntax: `60s`.

Full stack trace:
```
2024-07-05 13:31:32,203 INFO  [io.qua.arc.pro.BeanProcessor] (main) Found unrecommended usage of private members (use package-private instead) in application beans:
	- @Inject field io.quarkus.issue.MyComponent#myDuration

java.lang.RuntimeException: Error injecting java.time.Duration io.quarkus.issue.MyComponent.myDuration

	at io.quarkus.issue.MyComponent_Bean.doCreate(Unknown Source)
	at io.quarkus.issue.MyComponent_Bean.create(Unknown Source)
	at io.quarkus.issue.MyComponent_Bean.create(Unknown Source)
	at io.quarkus.arc.impl.AbstractSharedContext.createInstanceHandle(AbstractSharedContext.java:119)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:38)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:35)
	at io.quarkus.arc.impl.LazyValue.get(LazyValue.java:32)
	at io.quarkus.arc.impl.ComputingCache.computeIfAbsent(ComputingCache.java:69)
	at io.quarkus.arc.impl.ComputingCacheContextInstances.computeIfAbsent(ComputingCacheContextInstances.java:19)
	at io.quarkus.arc.impl.AbstractSharedContext.get(AbstractSharedContext.java:35)
	at io.quarkus.arc.impl.ClientProxies.getApplicationScopedDelegate(ClientProxies.java:21)
	at io.quarkus.issue.MyComponent_ClientProxy.arc$delegate(Unknown Source)
	at io.quarkus.issue.MyComponent_ClientProxy.getMyDuration(Unknown Source)
	at io.quarkus.issue.MyComponentTest.testMyDuration(MyComponentTest.java:17)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1597)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1597)
Caused by: jakarta.enterprise.inject.CreationException: Error creating synthetic bean [WtUE35eXet8tKWS0XAkVy-_OpGY]: java.lang.IllegalArgumentException: SRCFG00039: The config property sco.maxcurrentlistener.down.duration with the config value "60s" threw an Exception whilst being converted SRCFG00020: Failed to convert value with static method
	at io.quarkus.arc.generator.Object_WtUE35eXet8tKWS0XAkVy-_OpGY_Synthetic_Bean.doCreate(Unknown Source)
	at io.quarkus.arc.generator.Object_WtUE35eXet8tKWS0XAkVy-_OpGY_Synthetic_Bean.create(Unknown Source)
	at io.quarkus.arc.generator.Object_WtUE35eXet8tKWS0XAkVy-_OpGY_Synthetic_Bean.get(Unknown Source)
	at io.quarkus.arc.impl.CurrentInjectionPointProvider.get(CurrentInjectionPointProvider.java:48)
	... 17 more
Caused by: java.lang.IllegalArgumentException: SRCFG00039: The config property sco.maxcurrentlistener.down.duration with the config value "60s" threw an Exception whilst being converted SRCFG00020: Failed to convert value with static method
	at io.smallrye.config.SmallRyeConfig.convertValue(SmallRyeConfig.java:421)
	at io.smallrye.config.inject.ConfigProducerUtil.getValue(ConfigProducerUtil.java:100)
	at io.smallrye.config.inject.ConfigProducerUtil.getValue(ConfigProducerUtil.java:60)
	at io.quarkus.test.component.ConfigPropertyBeanCreator.create(ConfigPropertyBeanCreator.java:40)
	at io.quarkus.arc.generator.Object_WtUE35eXet8tKWS0XAkVy-_OpGY_Synthetic_Bean.createSynthetic(Unknown Source)
	... 21 more
Caused by: java.lang.IllegalArgumentException: SRCFG00020: Failed to convert value with static method
	at io.smallrye.config.ImplicitConverters$StaticMethodConverter.convert(ImplicitConverters.java:133)
	at io.smallrye.config.SmallRyeConfig.convertValue(SmallRyeConfig.java:419)
	... 25 more
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:118)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at io.smallrye.config.ImplicitConverters$StaticMethodConverter.convert(ImplicitConverters.java:131)
	... 26 more
Caused by: java.time.format.DateTimeParseException: Text cannot be parsed to a Duration
	at java.base/java.time.Duration.parse(Duration.java:419)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	... 28 more


Process finished with exit code 255
```

### How to Reproduce?

Component:
```
package io.quarkus.issue;

import jakarta.enterprise.context.ApplicationScoped;
import org.eclipse.microprofile.config.inject.ConfigProperty;

import java.time.Duration;

@ApplicationScoped
public class MyComponent {

    @ConfigProperty(name = "sco.maxcurrentlistener.down.duration", defaultValue = "60s")
    private Duration myDuration;

    public String getMyDuration() {
        return myDuration.toString();
    }
}
```

Component Test:
```
package io.quarkus.issue;

import io.quarkus.test.component.QuarkusComponentTest;
import jakarta.inject.Inject;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

@QuarkusComponentTest
class MyComponentTest {

    @Inject
    MyComponent myComponent;

    @Test
    public void testMyDuration() {
        assertEquals("60s", myComponent.getMyDuration());
    }
}
```

### Output of `uname -a` or `ver`

Darwin MacBook-Pro-8.local 22.5.0 Darwin Kernel Version 22.5.0: Mon Apr 24 20:53:19 PDT 2023; root:xnu-8796.121.2~5/RELEASE_ARM64_T6020 arm64

### Output of `java -version`

openjdk version "22" 2024-03-19 OpenJDK Runtime Environment Temurin-22+36 (build 22+36) OpenJDK 64-Bit Server VM Temurin-22+36 (build 22+36, mixed mode)

### Quarkus version or git rev

3.11.1

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Maven home: /Users/user/.m2/wrapper/dists/apache-maven-3.9.3-bin/326f10f4/apache-maven-3.9.3 Java version: 22, vendor: Eclipse Adoptium, runtime: /Users/user/.sdkman/candidates/java/22-tem Default locale: en_GB, platform encoding: UTF-8 OS name: "mac os x", version: "13.4", arch: "aarch64", family: "mac"

### Additional information

I tried adding the converter manually in test code by
```java
    @BeforeEach
    public void registerConverter() {
        var instance = ConfigProviderResolver.instance();
        var builder = instance.getBuilder()
                .withConverters(new DurationConverter())
                .build();
        instance.registerConfig(builder, MyComponentTest.class.getClassLoader());
    }
```
or adding in the test method itself but this does not work.
The actual used `Config` from which the `Converter` is obtained is different than the one created in the test.
When trying to use the correct ClassLoader
```java
        var contextClassLoader = Thread.currentThread().getContextClassLoader();
        instance.registerConfig(builder, contextClassLoader);
```
the test failed since a `Config` is already registered for that class loader.

--- a/test-framework/junit5-component/src/main/java/io/quarkus/test/component/QuarkusComponentTest.java
+++ b/test-framework/junit5-component/src/main/java/io/quarkus/test/component/QuarkusComponentTest.java
@@ -6,6 +6,7 @@ import static java.lang.annotation.RetentionPolicy.RUNTIME;
 import java.lang.annotation.Retention;
 import java.lang.annotation.Target;
 
+import org.eclipse.microprofile.config.spi.Converter;
 import org.junit.jupiter.api.extension.ExtendWith;
 
 import io.quarkus.arc.processor.AnnotationsTransformer;
@@ -71,5 +72,12 @@ public @interface QuarkusComponentTest {
      * @see QuarkusComponentTestExtensionBuilder#addAnnotationsTransformer(AnnotationsTransformer)
      */
     Class<? extends AnnotationsTransformer>[] annotationsTransformers() default {};
+
+    /**
+     * The additional config converters. By default, the Quarkus-specific converters are registered.
+     *
+     * @see QuarkusComponentTestExtensionBuilder#addConverter(Converter)
+     */
+    Class<? extends Converter<?>>[] configConverters() default {};
 
 }

--- a/test-framework/junit5-component/src/main/java/io/quarkus/test/component/QuarkusComponentTestConfiguration.java
+++ b/test-framework/junit5-component/src/main/java/io/quarkus/test/component/QuarkusComponentTestConfiguration.java
@@ -9,6 +9,7 @@ import java.util.Collections;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.function.Consumer;
 
 import jakarta.enterprise.event.Event;
 import jakarta.enterprise.inject.Instance;
@@ -34,6 +35,7 @@ import io.quarkus.runtime.configuration.RegexConverter;
 import io.quarkus.runtime.configuration.ZoneIdConverter;
 import io.quarkus.runtime.logging.LevelConverter;
 import io.quarkus.test.InjectMock;
+import io.smallrye.config.SmallRyeConfigBuilder;
 
 class QuarkusComponentTestConfiguration {
 
@@ -52,7 +54,7 @@ class QuarkusComponentTestConfiguration {
 
     static final QuarkusComponentTestConfiguration DEFAULT = new QuarkusComponentTestConfiguration(Map.of(), List.of(),
             List.of(), false, true, QuarkusComponentTestExtensionBuilder.DEFAULT_CONFIG_SOURCE_ORDINAL, List.of(),
-            DEFAULT_CONVERTERS);
+            DEFAULT_CONVERTERS, null);
 
     private static final Logger LOG = Logger.getLogger(QuarkusComponentTestConfiguration.class);
 
@@ -64,11 +66,13 @@ class QuarkusComponentTestConfiguration {
     final int configSourceOrdinal;
     final List<AnnotationsTransformer> annotationsTransformers;
     final List<Converter<?>> configConverters;
+    final Consumer<SmallRyeConfigBuilder> configBuilderCustomizer;
 
     QuarkusComponentTestConfiguration(Map<String, String> configProperties, List<Class<?>> componentClasses,
             List<MockBeanConfiguratorImpl<?>> mockConfigurators, boolean useDefaultConfigProperties,
             boolean addNestedClassesAsComponents, int configSourceOrdinal,
-            List<AnnotationsTransformer> annotationsTransformers, List<Converter<?>> configConverters) {
+            List<AnnotationsTransformer> annotationsTransformers, List<Converter<?>> configConverters,
+            Consumer<SmallRyeConfigBuilder> configBuilderCustomizer) {
         this.configProperties = configProperties;
         this.componentClasses = componentClasses;
         this.mockConfigurators = mockConfigurators;
@@ -77,6 +81,7 @@ class QuarkusComponentTestConfiguration {
         this.configSourceOrdinal = configSourceOrdinal;
         this.annotationsTransformers = annotationsTransformers;
         this.configConverters = configConverters;
+        this.configBuilderCustomizer = configBuilderCustomizer;
     }
 
     QuarkusComponentTestConfiguration update(Class<?> testClass) {
@@ -159,7 +164,7 @@ class QuarkusComponentTestConfiguration {
         return new QuarkusComponentTestConfiguration(Map.copyOf(configProperties), List.copyOf(componentClasses),
                 this.mockConfigurators,
                 useDefaultConfigProperties, addNestedClassesAsComponents, configSourceOrdinal,
-                List.copyOf(annotationsTransformers), List.copyOf(configConverters));
+                List.copyOf(annotationsTransformers), List.copyOf(configConverters), configBuilderCustomizer);
     }
 
     QuarkusComponentTestConfiguration update(Method testMethod) {
@@ -171,7 +176,7 @@ class QuarkusComponentTestConfiguration {
         }
         return new QuarkusComponentTestConfiguration(configProperties, componentClasses,
                 mockConfigurators, useDefaultConfigProperties, addNestedClassesAsComponents, configSourceOrdinal,
-                annotationsTransformers, configConverters);
+                annotationsTransformers, configConverters, configBuilderCustomizer);
     }
 
     private static boolean resolvesToBuiltinBean(Class<?> rawType) {

--- a/test-framework/junit5-component/src/main/java/io/quarkus/test/component/QuarkusComponentTestConfiguration.java
+++ b/test-framework/junit5-component/src/main/java/io/quarkus/test/component/QuarkusComponentTestConfiguration.java
@@ -17,16 +17,42 @@ import jakarta.enterprise.inject.spi.BeanManager;
 import jakarta.inject.Inject;
 import jakarta.inject.Provider;
 
+import org.eclipse.microprofile.config.spi.Converter;
 import org.jboss.logging.Logger;
 
 import io.quarkus.arc.InjectableInstance;
 import io.quarkus.arc.processor.AnnotationsTransformer;
+import io.quarkus.runtime.configuration.CharsetConverter;
+import io.quarkus.runtime.configuration.CidrAddressConverter;
+import io.quarkus.runtime.configuration.DurationConverter;
+import io.quarkus.runtime.configuration.InetAddressConverter;
+import io.quarkus.runtime.configuration.InetSocketAddressConverter;
+import io.quarkus.runtime.configuration.LocaleConverter;
+import io.quarkus.runtime.configuration.MemorySizeConverter;
+import io.quarkus.runtime.configuration.PathConverter;
+import io.quarkus.runtime.configuration.RegexConverter;
+import io.quarkus.runtime.configuration.ZoneIdConverter;
+import io.quarkus.runtime.logging.LevelConverter;
 import io.quarkus.test.InjectMock;
 
 class QuarkusComponentTestConfiguration {
 
+    // As defined in /quarkus/core/runtime/src/main/resources/META-INF/services/org.eclipse.microprofile.config.spi.Converter
+    static final List<Converter<?>> DEFAULT_CONVERTERS = List.of(new InetSocketAddressConverter(),
+            new CharsetConverter(),
+            new CidrAddressConverter(),
+            new InetAddressConverter(),
+            new RegexConverter(),
+            new PathConverter(),
+            new DurationConverter(),
+            new MemorySizeConverter(),
+            new LocaleConverter(),
+            new ZoneIdConverter(),
+            new LevelConverter());
+
     static final QuarkusComponentTestConfiguration DEFAULT = new QuarkusComponentTestConfiguration(Map.of(), List.of(),
-            List.of(), false, true, QuarkusComponentTestExtensionBuilder.DEFAULT_CONFIG_SOURCE_ORDINAL, List.of());
+            List.of(), false, true, QuarkusComponentTestExtensionBuilder.DEFAULT_CONFIG_SOURCE_ORDINAL, List.of(),
+            DEFAULT_CONVERTERS);
 
     private static final Logger LOG = Logger.getLogger(QuarkusComponentTestConfiguration.class);
 
@@ -37,11 +63,12 @@ class QuarkusComponentTestConfiguration {
     final boolean addNestedClassesAsComponents;
     final int configSourceOrdinal;
     final List<AnnotationsTransformer> annotationsTransformers;
+    final List<Converter<?>> configConverters;
 
     QuarkusComponentTestConfiguration(Map<String, String> configProperties, List<Class<?>> componentClasses,
             List<MockBeanConfiguratorImpl<?>> mockConfigurators, boolean useDefaultConfigProperties,
             boolean addNestedClassesAsComponents, int configSourceOrdinal,
-            List<AnnotationsTransformer> annotationsTransformers) {
+            List<AnnotationsTransformer> annotationsTransformers, List<Converter<?>> configConverters) {
         this.configProperties = configProperties;
         this.componentClasses = componentClasses;
         this.mockConfigurators = mockConfigurators;
@@ -49,6 +76,7 @@ class QuarkusComponentTestConfiguration {
         this.addNestedClassesAsComponents = addNestedClassesAsComponents;
         this.configSourceOrdinal = configSourceOrdinal;
         this.annotationsTransformers = annotationsTransformers;
+        this.configConverters = configConverters;
     }
 
     QuarkusComponentTestConfiguration update(Class<?> testClass) {
@@ -58,6 +86,7 @@ class QuarkusComponentTestConfiguration {
         boolean addNestedClassesAsComponents = this.addNestedClassesAsComponents;
         int configSourceOrdinal = this.configSourceOrdinal;
         List<AnnotationsTransformer> annotationsTransformers = new ArrayList<>(this.annotationsTransformers);
+        List<Converter<?>> configConverters = new ArrayList<>(this.configConverters);
 
         QuarkusComponentTest testAnnotation = testClass.getAnnotation(QuarkusComponentTest.class);
         if (testAnnotation != null) {
@@ -71,7 +100,17 @@ class QuarkusComponentTestConfiguration {
                     try {
                         annotationsTransformers.add(transformerClass.getDeclaredConstructor().newInstance());
                     } catch (Exception e) {
-                        LOG.errorf("Unable to instantiate %s", transformerClass);
+                        LOG.errorf(e, "Unable to instantiate %s", transformerClass);
+                    }
+                }
+            }
+            Class<? extends Converter<?>>[] converters = testAnnotation.configConverters();
+            if (converters.length > 0) {
+                for (Class<? extends Converter<?>> converterClass : converters) {
+                    try {
+                        configConverters.add(converterClass.getDeclaredConstructor().newInstance());
+                    } catch (Exception e) {
+                        LOG.errorf(e, "Unable to instantiate %s", converterClass);
                     }
                 }
             }
@@ -120,7 +159,7 @@ class QuarkusComponentTestConfiguration {
         return new QuarkusComponentTestConfiguration(Map.copyOf(configProperties), List.copyOf(componentClasses),
                 this.mockConfigurators,
                 useDefaultConfigProperties, addNestedClassesAsComponents, configSourceOrdinal,
-                List.copyOf(annotationsTransformers));
+                List.copyOf(annotationsTransformers), List.copyOf(configConverters));
     }
 
     QuarkusComponentTestConfiguration update(Method testMethod) {
@@ -132,7 +171,7 @@ class QuarkusComponentTestConfiguration {
         }
         return new QuarkusComponentTestConfiguration(configProperties, componentClasses,
                 mockConfigurators, useDefaultConfigProperties, addNestedClassesAsComponents, configSourceOrdinal,
-                annotationsTransformers);
+                annotationsTransformers, configConverters);
     }
 
     private static boolean resolvesToBuiltinBean(Class<?> rawType) {

--- a/test-framework/junit5-component/src/main/java/io/quarkus/test/component/QuarkusComponentTestExtension.java
+++ b/test-framework/junit5-component/src/main/java/io/quarkus/test/component/QuarkusComponentTestExtension.java
@@ -79,6 +79,7 @@ import org.junit.jupiter.api.extension.AfterEachCallback;
 import org.junit.jupiter.api.extension.BeforeAllCallback;
 import org.junit.jupiter.api.extension.BeforeEachCallback;
 import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.ExtensionContext.Store;
 import org.junit.jupiter.api.extension.ParameterContext;
 import org.junit.jupiter.api.extension.ParameterResolutionException;
 import org.junit.jupiter.api.extension.ParameterResolver;
@@ -204,7 +205,7 @@ public class QuarkusComponentTestExtension
     public QuarkusComponentTestExtension(Class<?>... additionalComponentClasses) {
         this(new QuarkusComponentTestConfiguration(Map.of(), List.of(additionalComponentClasses),
                 List.of(), false, true, QuarkusComponentTestExtensionBuilder.DEFAULT_CONFIG_SOURCE_ORDINAL,
-                List.of(), List.of()));
+                List.of(), List.of(), null));
     }
 
     QuarkusComponentTestExtension(QuarkusComponentTestConfiguration baseConfiguration) {
@@ -252,7 +253,7 @@ public class QuarkusComponentTestExtension
     @Override
     public void postProcessTestInstance(Object testInstance, ExtensionContext context) throws Exception {
         long start = System.nanoTime();
-        context.getRoot().getStore(NAMESPACE).put(KEY_TEST_INSTANCE, testInstance);
+        store(context).put(KEY_TEST_INSTANCE, testInstance);
         LOG.debugf("postProcessTestInstance: %s ms", TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - start));
     }
 
@@ -322,7 +323,7 @@ public class QuarkusComponentTestExtension
     public Object resolveParameter(ParameterContext parameterContext, ExtensionContext context)
             throws ParameterResolutionException {
         @SuppressWarnings("unchecked")
-        List<InstanceHandle<?>> injectedParams = context.getRoot().getStore(NAMESPACE).get(KEY_INJECTED_PARAMS, List.class);
+        List<InstanceHandle<?>> injectedParams = store(context).get(KEY_INJECTED_PARAMS, List.class);
         ArcContainer container = Arc.container();
         BeanManager beanManager = container.beanManager();
         java.lang.reflect.Type requiredType = parameterContext.getParameter().getParameterizedType();
@@ -339,7 +340,7 @@ public class QuarkusComponentTestExtension
 
     private void destroyDependentTestMethodParams(ExtensionContext context) {
         @SuppressWarnings("unchecked")
-        List<InstanceHandle<?>> injectedParams = context.getRoot().getStore(NAMESPACE).get(KEY_INJECTED_PARAMS, List.class);
+        List<InstanceHandle<?>> injectedParams = store(context).get(KEY_INJECTED_PARAMS, List.class);
         for (InstanceHandle<?> handle : injectedParams) {
             if (handle.getBean() != null && handle.getBean().getScope().equals(Dependent.class)) {
                 try {
@@ -355,17 +356,17 @@ public class QuarkusComponentTestExtension
     private void buildContainer(ExtensionContext context) {
         QuarkusComponentTestConfiguration testClassConfiguration = baseConfiguration
                 .update(context.getRequiredTestClass());
-        context.getRoot().getStore(NAMESPACE).put(KEY_TEST_CLASS_CONFIG, testClassConfiguration);
+        store(context).put(KEY_TEST_CLASS_CONFIG, testClassConfiguration);
         ClassLoader oldTccl = initArcContainer(context, testClassConfiguration);
-        context.getRoot().getStore(NAMESPACE).put(KEY_OLD_TCCL, oldTccl);
+        store(context).put(KEY_OLD_TCCL, oldTccl);
     }
 
     @SuppressWarnings("unchecked")
     private void cleanup(ExtensionContext context) {
-        ClassLoader oldTccl = context.getRoot().getStore(NAMESPACE).get(KEY_OLD_TCCL, ClassLoader.class);
+        ClassLoader oldTccl = store(context).get(KEY_OLD_TCCL, ClassLoader.class);
         Thread.currentThread().setContextClassLoader(oldTccl);
-        context.getRoot().getStore(NAMESPACE).remove(KEY_CONFIG_MAPPINGS);
-        Set<Path> generatedResources = context.getRoot().getStore(NAMESPACE).get(KEY_GENERATED_RESOURCES, Set.class);
+        store(context).remove(KEY_CONFIG_MAPPINGS);
+        Set<Path> generatedResources = store(context).get(KEY_GENERATED_RESOURCES, Set.class);
         for (Path path : generatedResources) {
             try {
                 LOG.debugf("Delete generated %s", path);
@@ -379,7 +380,7 @@ public class QuarkusComponentTestExtension
     @SuppressWarnings("unchecked")
     private void stopContainer(ExtensionContext context, Lifecycle testInstanceLifecycle) throws Exception {
         if (testInstanceLifecycle.equals(context.getTestInstanceLifecycle().orElse(Lifecycle.PER_METHOD))) {
-            for (FieldInjector fieldInjector : (List<FieldInjector>) context.getRoot().getStore(NAMESPACE)
+            for (FieldInjector fieldInjector : (List<FieldInjector>) store(context)
                     .get(KEY_INJECTED_FIELDS, List.class)) {
                 fieldInjector.unset(context.getRequiredTestInstance());
             }
@@ -392,10 +393,10 @@ public class QuarkusComponentTestExtension
             ConfigBeanCreator.clear();
             InterceptorMethodCreator.clear();
 
-            SmallRyeConfig config = context.getRoot().getStore(NAMESPACE).get(KEY_CONFIG, SmallRyeConfig.class);
+            SmallRyeConfig config = store(context).get(KEY_CONFIG, SmallRyeConfig.class);
             ConfigProviderResolver.instance().releaseConfig(config);
             ConfigProviderResolver
-                    .setInstance(context.getRoot().getStore(NAMESPACE).get(KEY_OLD_CONFIG_PROVIDER_RESOLVER,
+                    .setInstance(store(context).get(KEY_OLD_CONFIG_PROVIDER_RESOLVER,
                             ConfigProviderResolver.class));
         }
     }
@@ -405,15 +406,15 @@ public class QuarkusComponentTestExtension
             // Init ArC
             Arc.initialize();
 
-            QuarkusComponentTestConfiguration configuration = context.getRoot().getStore(NAMESPACE)
-                    .get(KEY_TEST_CLASS_CONFIG, QuarkusComponentTestConfiguration.class);
+            QuarkusComponentTestConfiguration configuration = store(context).get(KEY_TEST_CLASS_CONFIG,
+                    QuarkusComponentTestConfiguration.class);
             Optional<Method> testMethod = context.getTestMethod();
             if (testMethod.isPresent()) {
                 configuration = configuration.update(testMethod.get());
             }
 
             ConfigProviderResolver oldConfigProviderResolver = ConfigProviderResolver.instance();
-            context.getRoot().getStore(NAMESPACE).put(KEY_OLD_CONFIG_PROVIDER_RESOLVER, oldConfigProviderResolver);
+            store(context).put(KEY_OLD_CONFIG_PROVIDER_RESOLVER, oldConfigProviderResolver);
 
             SmallRyeConfigProviderResolver smallRyeConfigProviderResolver = new SmallRyeConfigProviderResolver();
             ConfigProviderResolver.setInstance(smallRyeConfigProviderResolver);
@@ -430,26 +431,31 @@ public class QuarkusComponentTestExtension
                             new QuarkusComponentTestConfigSource(configuration.configProperties,
                                     configuration.configSourceOrdinal));
             @SuppressWarnings("unchecked")
-            Set<ConfigClassWithPrefix> configMappings = context.getRoot().getStore(NAMESPACE).get(KEY_CONFIG_MAPPINGS,
-                    Set.class);
+            Set<ConfigClassWithPrefix> configMappings = store(context).get(KEY_CONFIG_MAPPINGS, Set.class);
             if (configMappings != null) {
                 // Register the mappings found during bean discovery
                 for (ConfigClassWithPrefix mapping : configMappings) {
                     configBuilder.withMapping(mapping.getKlass(), mapping.getPrefix());
                 }
             }
+            if (configuration.configBuilderCustomizer != null) {
+                configuration.configBuilderCustomizer.accept(configBuilder);
+            }
             SmallRyeConfig config = configBuilder.build();
             smallRyeConfigProviderResolver.registerConfig(config, tccl);
-            context.getRoot().getStore(NAMESPACE).put(KEY_CONFIG, config);
+            store(context).put(KEY_CONFIG, config);
             ConfigBeanCreator.setClassLoader(tccl);
 
             // Inject fields declated on the test class
             Object testInstance = context.getRequiredTestInstance();
-            context.getRoot().getStore(NAMESPACE).put(KEY_INJECTED_FIELDS,
-                    injectFields(context.getRequiredTestClass(), testInstance));
+            store(context).put(KEY_INJECTED_FIELDS, injectFields(context.getRequiredTestClass(), testInstance));
             // Injected test method parameters
-            context.getRoot().getStore(NAMESPACE).put(KEY_INJECTED_PARAMS, new CopyOnWriteArrayList<>());
+            store(context).put(KEY_INJECTED_PARAMS, new CopyOnWriteArrayList<>());
         }
+    }
+
+    private Store store(ExtensionContext context) {
+        return context.getRoot().getStore(NAMESPACE);
     }
 
     private BeanRegistrar registrarForMock(MockBeanConfiguratorImpl<?> mock) {
@@ -633,7 +639,7 @@ public class QuarkusComponentTestExtension
                 });
             }
 
-            extensionContext.getRoot().getStore(NAMESPACE).put(KEY_GENERATED_RESOURCES, generatedResources);
+            store(extensionContext).put(KEY_GENERATED_RESOURCES, generatedResources);
 
             builder.addAnnotationTransformation(AnnotationsTransformer.appliedToField().whenContainsAny(qualifiers)
                     .whenContainsNone(DotName.createSimple(Inject.class)).thenTransform(t -> t.add(Inject.class)));
@@ -783,7 +789,7 @@ public class QuarkusComponentTestExtension
                                         .configClassWithPrefix(ConfigMappingBeanCreator.tryLoad(mapping), e.getKey()));
                             }
                         }
-                        extensionContext.getRoot().getStore(NAMESPACE).put(KEY_CONFIG_MAPPINGS, configMappings);
+                        store(extensionContext).put(KEY_CONFIG_MAPPINGS, configMappings);
                     }
 
                     LOG.debugf("Test injection points analyzed in %s ms [found: %s, mocked: %s]",
@@ -850,7 +856,7 @@ public class QuarkusComponentTestExtension
         return oldTccl;
     }
 
-    private void processTestInterceptorMethods(Class<?> testClass, ExtensionContext extensionContext,
+    private void processTestInterceptorMethods(Class<?> testClass, ExtensionContext context,
             BeanRegistrar.RegistrationContext registrationContext, Set<String> interceptorBindings) {
         List<Class<? extends Annotation>> annotations = List.of(AroundInvoke.class, PostConstruct.class, PreDestroy.class,
                 AroundConstruct.class);
@@ -873,7 +879,7 @@ public class QuarkusComponentTestExtension
                 return ic -> {
                     Object instance = null;
                     if (!Modifier.isStatic(method.getModifiers())) {
-                        Object testInstance = extensionContext.getRoot().getStore(NAMESPACE).get(KEY_TEST_INSTANCE);
+                        Object testInstance = store(context).get(KEY_TEST_INSTANCE);
                         if (testInstance == null) {
                             throw new IllegalStateException("Test instance not available");
                         }

--- a/test-framework/junit5-component/src/main/java/io/quarkus/test/component/QuarkusComponentTestExtension.java
+++ b/test-framework/junit5-component/src/main/java/io/quarkus/test/component/QuarkusComponentTestExtension.java
@@ -57,6 +57,7 @@ import jakarta.interceptor.InvocationContext;
 import org.eclipse.microprofile.config.Config;
 import org.eclipse.microprofile.config.inject.ConfigProperty;
 import org.eclipse.microprofile.config.spi.ConfigProviderResolver;
+import org.eclipse.microprofile.config.spi.Converter;
 import org.jboss.jandex.AnnotationInstance;
 import org.jboss.jandex.AnnotationValue;
 import org.jboss.jandex.ClassInfo;
@@ -203,7 +204,7 @@ public class QuarkusComponentTestExtension
     public QuarkusComponentTestExtension(Class<?>... additionalComponentClasses) {
         this(new QuarkusComponentTestConfiguration(Map.of(), List.of(additionalComponentClasses),
                 List.of(), false, true, QuarkusComponentTestExtensionBuilder.DEFAULT_CONFIG_SOURCE_ORDINAL,
-                List.of()));
+                List.of(), List.of()));
     }
 
     QuarkusComponentTestExtension(QuarkusComponentTestConfiguration baseConfiguration) {
@@ -421,6 +422,7 @@ public class QuarkusComponentTestExtension
             ClassLoader tccl = Thread.currentThread().getContextClassLoader();
             SmallRyeConfigBuilder configBuilder = new SmallRyeConfigBuilder().forClassLoader(tccl)
                     .addDefaultInterceptors()
+                    .withConverters(configuration.configConverters.toArray(new Converter<?>[] {}))
                     .addDefaultSources()
                     .withSources(new ApplicationPropertiesConfigSourceLoader.InFileSystem())
                     .withSources(new ApplicationPropertiesConfigSourceLoader.InClassPath())

--- a/test-framework/junit5-component/src/main/java/io/quarkus/test/component/QuarkusComponentTestExtensionBuilder.java
+++ b/test-framework/junit5-component/src/main/java/io/quarkus/test/component/QuarkusComponentTestExtensionBuilder.java
@@ -5,11 +5,13 @@ import java.util.Collections;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.function.Consumer;
 import java.util.function.Function;
 
 import org.eclipse.microprofile.config.spi.Converter;
 
 import io.quarkus.arc.processor.AnnotationsTransformer;
+import io.smallrye.config.SmallRyeConfigBuilder;
 
 /**
  * Convenient builder for {@link QuarkusComponentTestExtension}.
@@ -32,6 +34,7 @@ public class QuarkusComponentTestExtensionBuilder {
     private boolean useDefaultConfigProperties = false;
     private boolean addNestedClassesAsComponents = true;
     private int configSourceOrdinal = QuarkusComponentTestExtensionBuilder.DEFAULT_CONFIG_SOURCE_ORDINAL;
+    private Consumer<SmallRyeConfigBuilder> configBuilderCustomizer;
 
     /**
      * The initial set of components under test is derived from the test class. The types of all fields annotated with
@@ -111,11 +114,25 @@ public class QuarkusComponentTestExtensionBuilder {
     /**
      * Add an additional {@link Converter}. By default, the Quarkus-specific converters are registered.
      *
-     * @param transformer
+     * @param converter
      * @return self
+     * @see #setConfigBuilderCustomizer(Consumer)
      */
     public QuarkusComponentTestExtensionBuilder addConverter(Converter<?> converter) {
         configConverters.add(converter);
+        return this;
+    }
+
+    /**
+     * Set the {@link SmallRyeConfigBuilder} customizer.
+     * <p>
+     * The customizer can affect the configuration of a test method and should be used with caution.
+     *
+     * @param customizer
+     * @return self
+     */
+    public QuarkusComponentTestExtensionBuilder setConfigBuilderCustomizer(Consumer<SmallRyeConfigBuilder> customizer) {
+        this.configBuilderCustomizer = customizer;
         return this;
     }
 
@@ -149,7 +166,7 @@ public class QuarkusComponentTestExtensionBuilder {
         return new QuarkusComponentTestExtension(new QuarkusComponentTestConfiguration(Map.copyOf(configProperties),
                 List.copyOf(componentClasses), List.copyOf(mockConfigurators), useDefaultConfigProperties,
                 addNestedClassesAsComponents, configSourceOrdinal,
-                List.copyOf(annotationsTransformers), converters));
+                List.copyOf(annotationsTransformers), converters, configBuilderCustomizer));
     }
 
     void registerMockBean(MockBeanConfiguratorImpl<?> mock) {

--- a/test-framework/junit5-component/src/main/java/io/quarkus/test/component/QuarkusComponentTestExtensionBuilder.java
+++ b/test-framework/junit5-component/src/main/java/io/quarkus/test/component/QuarkusComponentTestExtensionBuilder.java
@@ -7,6 +7,8 @@ import java.util.List;
 import java.util.Map;
 import java.util.function.Function;
 
+import org.eclipse.microprofile.config.spi.Converter;
+
 import io.quarkus.arc.processor.AnnotationsTransformer;
 
 /**
@@ -26,6 +28,7 @@ public class QuarkusComponentTestExtensionBuilder {
     private final List<Class<?>> componentClasses = new ArrayList<>();
     private final List<MockBeanConfiguratorImpl<?>> mockConfigurators = new ArrayList<>();
     private final List<AnnotationsTransformer> annotationsTransformers = new ArrayList<>();
+    private final List<Converter<?>> configConverters = new ArrayList<>();
     private boolean useDefaultConfigProperties = false;
     private boolean addNestedClassesAsComponents = true;
     private int configSourceOrdinal = QuarkusComponentTestExtensionBuilder.DEFAULT_CONFIG_SOURCE_ORDINAL;
@@ -106,6 +109,17 @@ public class QuarkusComponentTestExtensionBuilder {
     }
 
     /**
+     * Add an additional {@link Converter}. By default, the Quarkus-specific converters are registered.
+     *
+     * @param transformer
+     * @return self
+     */
+    public QuarkusComponentTestExtensionBuilder addConverter(Converter<?> converter) {
+        configConverters.add(converter);
+        return this;
+    }
+
+    /**
      * Configure a new mock of a bean.
      * <p>
      * Note that a mock is created automatically for all unsatisfied dependencies in the test. This API provides full control
@@ -124,10 +138,18 @@ public class QuarkusComponentTestExtensionBuilder {
      * @return a new extension instance
      */
     public QuarkusComponentTestExtension build() {
+        List<Converter<?>> converters;
+        if (configConverters.isEmpty()) {
+            converters = QuarkusComponentTestConfiguration.DEFAULT_CONVERTERS;
+        } else {
+            converters = new ArrayList<>(QuarkusComponentTestConfiguration.DEFAULT_CONVERTERS);
+            converters.addAll(configConverters);
+            converters = List.copyOf(converters);
+        }
         return new QuarkusComponentTestExtension(new QuarkusComponentTestConfiguration(Map.copyOf(configProperties),
-                List.copyOf(componentClasses),
-                List.copyOf(mockConfigurators), useDefaultConfigProperties, addNestedClassesAsComponents, configSourceOrdinal,
-                List.copyOf(annotationsTransformers)));
+                List.copyOf(componentClasses), List.copyOf(mockConfigurators), useDefaultConfigProperties,
+                addNestedClassesAsComponents, configSourceOrdinal,
+                List.copyOf(annotationsTransformers), converters));
     }
 
     void registerMockBean(MockBeanConfiguratorImpl<?> mock) {

--- a/test-framework/junit5-component/src/test/java/io/quarkus/test/component/config/ConfigBuilderCustomizerTest.java
+++ b/test-framework/junit5-component/src/test/java/io/quarkus/test/component/config/ConfigBuilderCustomizerTest.java
@@ -1,12 +1,9 @@
 package io.quarkus.test.component.config;
 
-import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 
-import java.time.Duration;
-import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
 
-import jakarta.annotation.Priority;
 import jakarta.inject.Singleton;
 
 import org.eclipse.microprofile.config.inject.ConfigProperty;
@@ -16,26 +13,27 @@ import org.junit.jupiter.api.extension.RegisterExtension;
 
 import io.quarkus.test.component.QuarkusComponentTestExtension;
 import io.quarkus.test.component.TestConfigProperty;
+import io.smallrye.config.SmallRyeConfigBuilder;
 
-public class ConfigConverterExtensionTest {
+public class ConfigBuilderCustomizerTest {
 
     @RegisterExtension
     static final QuarkusComponentTestExtension extension = QuarkusComponentTestExtension.builder()
-            .addConverter(new CustomBooleanConverter()).build();
+            .setConfigBuilderCustomizer(new Consumer<SmallRyeConfigBuilder>() {
+                @Override
+                public void accept(SmallRyeConfigBuilder builder) {
+                    builder.withConverter(Boolean.class, 300, new CustomBooleanConverter());
+                }
+            }).build();
 
     @TestConfigProperty(key = "my.boolean", value = "jo")
-    @TestConfigProperty(key = "my.duration", value = "5s")
     @Test
-    public void testConverters(Foo foo) {
-        assertEquals(TimeUnit.SECONDS.toMillis(5), foo.durationVal.toMillis());
+    public void testBuilderCustomizer(Foo foo) {
         assertTrue(foo.boolVal);
     }
 
     @Singleton
     public static class Foo {
-
-        @ConfigProperty(name = "my.duration", defaultValue = "60s")
-        Duration durationVal;
 
         @ConfigProperty(name = "my.boolean")
         boolean boolVal;
@@ -43,7 +41,6 @@ public class ConfigConverterExtensionTest {
     }
 
     @SuppressWarnings("serial")
-    @Priority(300)
     public static class CustomBooleanConverter implements Converter<Boolean> {
 
         @Override

--- a/test-framework/junit5-component/src/test/java/io/quarkus/test/component/config/ConfigConverterExtensionTest.java
+++ b/test-framework/junit5-component/src/test/java/io/quarkus/test/component/config/ConfigConverterExtensionTest.java
@@ -1,0 +1,55 @@
+package io.quarkus.test.component.config;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import java.time.Duration;
+import java.util.concurrent.TimeUnit;
+
+import jakarta.annotation.Priority;
+import jakarta.inject.Singleton;
+
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+import org.eclipse.microprofile.config.spi.Converter;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.component.QuarkusComponentTestExtension;
+import io.quarkus.test.component.TestConfigProperty;
+
+public class ConfigConverterExtensionTest {
+
+    @RegisterExtension
+    static final QuarkusComponentTestExtension extension = QuarkusComponentTestExtension.builder()
+            .addConverter(new CustomBooleanConverter()).build();
+
+    @TestConfigProperty(key = "my.boolean", value = "jo")
+    @TestConfigProperty(key = "my.duration", value = "5s")
+    @Test
+    public void testQuarkusDurationConverter(Foo foo) {
+        assertEquals(TimeUnit.SECONDS.toMillis(5), foo.durationVal.toMillis());
+        assertTrue(foo.boolVal);
+    }
+
+    @Singleton
+    public static class Foo {
+
+        @ConfigProperty(name = "my.duration", defaultValue = "60s")
+        Duration durationVal;
+
+        @ConfigProperty(name = "my.boolean")
+        boolean boolVal;
+
+    }
+
+    @SuppressWarnings("serial")
+    @Priority(300)
+    public static class CustomBooleanConverter implements Converter<Boolean> {
+
+        @Override
+        public Boolean convert(String value) throws IllegalArgumentException, NullPointerException {
+            return "jo".equals(value) ? true : false;
+        }
+
+    }
+}

--- a/test-framework/junit5-component/src/test/java/io/quarkus/test/component/config/ConfigConverterTest.java
+++ b/test-framework/junit5-component/src/test/java/io/quarkus/test/component/config/ConfigConverterTest.java
@@ -1,0 +1,52 @@
+package io.quarkus.test.component.config;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import java.time.Duration;
+import java.util.concurrent.TimeUnit;
+
+import jakarta.annotation.Priority;
+import jakarta.inject.Singleton;
+
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+import org.eclipse.microprofile.config.spi.Converter;
+import org.junit.jupiter.api.Test;
+
+import io.quarkus.test.component.QuarkusComponentTest;
+import io.quarkus.test.component.TestConfigProperty;
+import io.quarkus.test.component.config.ConfigConverterTest.CustomBooleanConverter;
+
+@QuarkusComponentTest(configConverters = CustomBooleanConverter.class)
+public class ConfigConverterTest {
+
+    @TestConfigProperty(key = "my.boolean", value = "jo")
+    @TestConfigProperty(key = "my.duration", value = "5s")
+    @Test
+    public void testQuarkusDurationConverter(Foo foo) {
+        assertEquals(TimeUnit.SECONDS.toMillis(5), foo.durationVal.toMillis());
+        assertTrue(foo.boolVal);
+    }
+
+    @Singleton
+    public static class Foo {
+
+        @ConfigProperty(name = "my.duration", defaultValue = "60s")
+        Duration durationVal;
+
+        @ConfigProperty(name = "my.boolean")
+        boolean boolVal;
+
+    }
+
+    @SuppressWarnings("serial")
+    @Priority(300)
+    public static class CustomBooleanConverter implements Converter<Boolean> {
+
+        @Override
+        public Boolean convert(String value) throws IllegalArgumentException, NullPointerException {
+            return "jo".equals(value) ? true : false;
+        }
+
+    }
+}

--- a/test-framework/junit5-component/src/test/java/io/quarkus/test/component/config/ConfigConverterTest.java
+++ b/test-framework/junit5-component/src/test/java/io/quarkus/test/component/config/ConfigConverterTest.java
@@ -23,7 +23,7 @@ public class ConfigConverterTest {
     @TestConfigProperty(key = "my.boolean", value = "jo")
     @TestConfigProperty(key = "my.duration", value = "5s")
     @Test
-    public void testQuarkusDurationConverter(Foo foo) {
+    public void testConverters(Foo foo) {
         assertEquals(TimeUnit.SECONDS.toMillis(5), foo.durationVal.toMillis());
         assertTrue(foo.boolVal);
     }

